src/push-state-tree.js

```diff
@@ -56,6 +51,59 @@ const CHANGE = 'change';
 const MATCH = 'match';
 const OLD_MATCH = 'oldMatch';
 
+// Internal history keep tracking of all changes in the location history, it can be reset any
+let internalHistory;
+function InternalLocation(id, url) {
+  this.id = id;
+  this.url = url;
+}
+function InternalHistory() {
+  if (!(this instanceof InternalHistory)) {
+    return internalHistory = new InternalHistory();
+  }
+
+  // It's a object like an array, but it can start from a certain point to allow memory cleanup
+  if (internalHistory && internalHistory.length) {
+    this.startFrom = internalHistory.length - 1;
+    let savedLocation = internalHistory[this.startFrom];
+
+    // Break the link to previous to allow GC collect unused values
+    if (savedLocation) savedLocation.previous = undefined;
+
+    this[this.startFrom] = savedLocation;
+    this.length = internalHistory.length;
```

Comment:
On a side note, is there a chance this class can be made to behave as an ordinary instance instead of a singleton-esque?